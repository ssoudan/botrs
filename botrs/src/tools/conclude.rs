use std::cell::RefCell;
use std::fmt::Debug;

use llm_chain::tools::{Describe, Format, Tool, ToolDescription, ToolUseError};
use serde::{Deserialize, Serialize};

use crate::tools::{TerminalTool, TerminationMessage};

/// A tool that is called to wrap the task.
#[derive(Default)]
pub struct ConcludeTool {
    done: RefCell<Option<ConcludeToolInput>>,
}

impl TerminalTool for ConcludeTool {
    fn is_done(&self) -> bool {
        self.done.borrow().is_some()
    }

    fn take_done(&self) -> Option<TerminationMessage> {
        self.done.borrow_mut().take().map(|input| input.into())
    }
}

/// A tool that is called to wrap the task.
#[derive(Debug, Serialize, Deserialize)]
pub struct ConcludeToolInput {
    /// The final textual answer for this task.
    pub conclusion: String,
    /// The original question that was asked to the user.
    pub original_question: String,
}

impl From<ConcludeToolInput> for TerminationMessage {
    fn from(input: ConcludeToolInput) -> Self {
        Self {
            conclusion: input.conclusion,
            original_question: input.original_question,
        }
    }
}

/// ConcludeToolOutput - empty
#[derive(Serialize, Deserialize)]
pub struct ConcludeToolOutput {}

impl Describe for ConcludeToolInput {
    fn describe() -> Format {
        vec![(
            "conclusion",
            "The final textual answer for this task. No string interpolation supported, only plain text. MANDATORY.",
            )
            .into(),
            (
                "original_question",
                "The original question that was asked to the user. No string interpolation supported, only plain text. MANDATORY.",
            ).into(),]
        .into()
    }
}

impl Describe for ConcludeToolOutput {
    fn describe() -> Format {
        vec![].into()
    }
}

impl ConcludeTool {
    fn invoke_typed(&self, input: ConcludeToolInput) -> Result<ConcludeToolOutput, ToolUseError> {
        if self.done.borrow().is_some() {
            return Err(ToolUseError::ToolInvocationFailed(
                "This task is already done.".to_string(),
            ));
        }

        *self.done.borrow_mut() = Some(input);

        Ok(ConcludeToolOutput {})
    }
}

impl Tool for ConcludeTool {
    fn description(&self) -> ToolDescription {
        ToolDescription::new(
            "Conclude",
            "A tool to conclude a task.",
            "You have to use this to once you have the answer to the task with your conclusion.",
            ConcludeToolInput::describe(),
            ConcludeToolOutput::describe(),
        )
    }

    fn invoke(&self, input: serde_yaml::Value) -> Result<serde_yaml::Value, ToolUseError> {
        let input = serde_yaml::from_value(input)?;
        let output = self.invoke_typed(input)?;
        Ok(serde_yaml::to_value(output)?)
    }
}
