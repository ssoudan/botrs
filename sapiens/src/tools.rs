use std::collections::HashMap;
use std::fmt::Debug;
use std::sync::Arc;

pub use llm_chain::parsing::{find_yaml, ExtractionError};
pub use llm_chain::tools::{Describe, Format, FormatPart, ToolDescription};
use serde::{Deserialize, Serialize};
use tokio::sync::RwLock;

/// Error while using a tool
#[derive(Debug, thiserror::Error)]
pub enum ToolUseError {
    /// Tool not found
    #[error("Tool not found: {0}")]
    ToolNotFound(String),
    /// Tool invocation failed
    #[error("Tool invocation failed: {0}")]
    ToolInvocationFailed(String),
    /// Invalid YAML
    #[error("Invalid YAML")]
    InvalidYaml(#[from] serde_yaml::Error),
    /// Invalid input
    #[error("Invalid input: {0}")]
    InvalidInput(#[from] ExtractionError),
}

#[derive(Serialize, Deserialize, Debug)]
struct ToolInvocationInput {
    command: String,
    input: serde_yaml::Value,
    output: Option<serde_yaml::Value>,
}

/// Something meant to become a [`Tool`] - description
pub trait ProtoToolDescribe {
    /// the description of the tool
    fn description(&self) -> ToolDescription;
}

/// Something meant to become a [`Tool`] - invocation
#[async_trait::async_trait]
pub trait ProtoToolInvoke {
    /// Invoke the tool
    async fn invoke(&self, input: serde_yaml::Value) -> Result<serde_yaml::Value, ToolUseError>;
}

/// A Tool - the most basic kind of tools. See [`AdvancedTool`] and
/// [`TerminalTool`] for more.
#[async_trait::async_trait]
pub trait Tool: Sync + Send {
    /// the description of the tool
    fn description(&self) -> ToolDescription;

    /// Invoke the tool
    async fn invoke(&self, input: serde_yaml::Value) -> Result<serde_yaml::Value, ToolUseError>;
}

#[async_trait::async_trait]
impl<T: Sync + Send> Tool for T
where
    T: ProtoToolDescribe + ProtoToolInvoke,
{
    fn description(&self) -> ToolDescription {
        self.description()
    }

    async fn invoke(&self, input: serde_yaml::Value) -> Result<serde_yaml::Value, ToolUseError> {
        self.invoke(input).await
    }
}

/// A termination message
///
/// This is the message that is sent to the user when a chain of exchanges
/// terminates.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TerminationMessage {
    /// The final textual answer for this task.
    pub conclusion: String,
    /// The original question that was asked to the user.
    pub original_question: String,
}

/// A [`Tool`] that wraps a chain of exchanges
#[async_trait::async_trait]
pub trait TerminalTool: Tool + Sync + Send {
    /// done flag.
    async fn is_done(&self) -> bool {
        false
    }

    /// Take the done flag.
    async fn take_done(&self) -> Option<TerminationMessage> {
        None
    }
}

/// A [`Tool`]  that can benefit from a [`Toolbox`]
#[async_trait::async_trait]
pub trait AdvancedTool: Tool {
    /// Invoke the tool with a [`Toolbox`]
    async fn invoke_with_toolbox(
        &self,
        toolbox: Toolbox,
        input: serde_yaml::Value,
    ) -> Result<serde_yaml::Value, ToolUseError>;
}

/// Toolbox
///
/// a [`Toolbox`] is a collection of [`Tool`], [`TerminalTool`] and
/// [`AdvancedTool`].
#[derive(Default, Clone)]
pub struct Toolbox {
    /// The terminal tools - the one that can terminate a chain of exchanges
    terminal_tools: Arc<RwLock<HashMap<String, Box<dyn TerminalTool>>>>,

    /// The tools - the other tools
    tools: Arc<RwLock<HashMap<String, Box<dyn Tool>>>>,

    /// The advanced tools - the one that can invoke another tool (not an
    /// advanced one)
    advanced_tools: Arc<RwLock<HashMap<String, Box<dyn AdvancedTool>>>>,
}

impl Debug for Toolbox {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Toolbox").finish()
    }
}

impl Toolbox {
    /// Collect the termination messages
    pub async fn termination_messages(&self) -> Vec<TerminationMessage> {
        let mut messages = Vec::new();

        for tool in self.terminal_tools.read().await.values() {
            if let Some(message) = tool.take_done().await {
                messages.push(message);
            }
        }

        messages
    }

    /// Add a terminal tool
    ///
    /// A [`TerminalTool`] can terminate a chain of exchanges.
    pub async fn add_terminal_tool(&mut self, tool: impl TerminalTool + 'static) {
        let name = tool.description().name;
        self.terminal_tools
            .write()
            .await
            .insert(name, Box::new(tool));
    }

    /// Add a tool
    ///
    /// A [`Tool`] can be invoked by an [`AdvancedTool`].
    pub async fn add_tool(&mut self, tool: impl Tool + 'static) {
        let name = tool.description().name;
        self.tools.write().await.insert(name, Box::new(tool));
    }

    /// Add an advanced tool
    ///
    /// An [`AdvancedTool`] is a [`Tool`] that can invoke another tool.
    pub async fn add_advanced_tool(&mut self, tool: impl AdvancedTool + 'static) {
        let name = tool.description().name;
        self.advanced_tools
            .write()
            .await
            .insert(name, Box::new(tool));
    }

    /// Get the descriptions of the tools
    pub async fn describe(&self) -> HashMap<String, ToolDescription> {
        let mut descriptions = HashMap::new();

        for (name, tool) in self.terminal_tools.read().await.iter() {
            descriptions.insert(name.clone(), tool.description());
        }

        for (name, tool) in self.tools.read().await.iter() {
            descriptions.insert(name.clone(), tool.description());
        }

        for (name, tool) in self.advanced_tools.read().await.iter() {
            descriptions.insert(name.clone(), tool.description());
        }

        descriptions
    }
}

/// Invoke a [`Tool`] or [`AdvancedTool`]  from a [`Toolbox`]
pub async fn invoke_from_toolbox(
    toolbox: Toolbox,
    name: &str,
    input: serde_yaml::Value,
) -> Result<serde_yaml::Value, ToolUseError> {
    // test if the tool is an advanced tool
    if let Some(tool) = toolbox.clone().advanced_tools.read().await.get(name) {
        return tool.invoke_with_toolbox(toolbox, input).await;
    }

    // if not, test if the tool is a terminal tool
    {
        let guard = toolbox.terminal_tools.read().await;
        if let Some(tool) = guard.get(name) {
            return tool.invoke(input).await;
        }
    }

    // otherwise, use the normal tool
    let guard = toolbox.tools.read().await;
    let tool = guard
        .get(name)
        .ok_or(ToolUseError::ToolNotFound(name.to_string()))?;

    tool.invoke(input).await
}

/// Invoke a Tool from a [`Toolbox`]
pub async fn invoke_simple_from_toolbox(
    toolbox: Toolbox,
    name: &str,
    input: serde_yaml::Value,
) -> Result<serde_yaml::Value, ToolUseError> {
    // test if the tool is a terminal tool
    {
        let guard = toolbox.terminal_tools.read().await;
        if let Some(tool) = guard.get(name) {
            return tool.invoke(input).await;
        }
    }

    // the normal tool only
    let guard = toolbox.tools.read().await;
    let tool = guard
        .get(name)
        .ok_or(ToolUseError::ToolNotFound(name.to_string()))?;

    tool.invoke(input).await
}

/// Try to find the tool invocation from the chat message and invoke the
/// corresponding tool.
///
/// If multiple tool invocations are found, only the first one is used.
#[tracing::instrument]
pub async fn invoke_tool(toolbox: Toolbox, data: &str) -> (String, Result<String, ToolUseError>) {
    let tool_invocations = find_yaml::<ToolInvocationInput>(data);

    match tool_invocations {
        Ok(tool_invocations) => {
            if tool_invocations.is_empty() {
                return (
                    "unknown".to_string(),
                    Err(ToolUseError::ToolInvocationFailed(
                        "No Action found".to_string(),
                    )),
                );
            }

            // if any tool_invocations have an 'output' field, we return an error
            for invocation in tool_invocations.iter() {
                if invocation.output.is_some() {
                    return (
                        "unknown".to_string(),
                        Err(ToolUseError::ToolInvocationFailed(
                            "The Action cannot have an `output` field. Only `command` and `input` are allowed.".to_string(),
                        )),
                    );
                }
            }

            // Take the first invocation - the list is reversed
            let invocation_input = &tool_invocations.last().unwrap();

            let tool_name = invocation_input.command.clone();

            let input = invocation_input.input.clone();

            match invoke_from_toolbox(toolbox, &invocation_input.command, input).await {
                Ok(o) => (tool_name, Ok(serde_yaml::to_string(&o).unwrap())),
                Err(e) => (tool_name, Err(e)),
            }
        }
        Err(e) => ("unknown".to_string(), Err(ToolUseError::InvalidInput(e))),
    }
}
