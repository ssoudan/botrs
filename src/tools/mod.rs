/// Bash tool
pub mod bash;
/// Hue tools
pub mod hue;
