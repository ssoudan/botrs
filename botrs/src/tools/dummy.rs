use std::fmt::Debug;

use llm_chain::tools::{Describe, Format, Tool, ToolDescription, ToolUseError};
use serde::{Deserialize, Serialize};

/// A tool that is called to test stuffs
#[derive(Default)]
pub struct DummyTool {}

/// A tool that is called to test stuffs
#[derive(Debug, Serialize, Deserialize)]
pub struct DummyToolInput {
    /// Well. MANDATORY.
    pub blah: String,
}

/// DummyToolOutput not very significant
#[derive(Serialize, Deserialize)]
pub struct DummyToolOutput {
    /// No much.
    pub something: String,
}

impl Describe for DummyToolInput {
    fn describe() -> Format {
        vec![("blah", "Well. MANDATORY.").into()].into()
    }
}

impl Describe for DummyToolOutput {
    fn describe() -> Format {
        vec![("something", "No much.").into()].into()
    }
}

impl DummyTool {
    fn invoke_typed(&self, input: &DummyToolInput) -> Result<DummyToolOutput, ToolUseError> {
        Ok(DummyToolOutput {
            something: input.blah.clone() + " and something else",
        })
    }
}

impl Tool for DummyTool {
    fn description(&self) -> ToolDescription {
        ToolDescription::new(
            "Dummy",
            "A tool to test stuffs.",
            "Use this to test stuffs.",
            DummyToolInput::describe(),
            DummyToolOutput::describe(),
        )
    }

    fn invoke(&self, input: serde_yaml::Value) -> Result<serde_yaml::Value, ToolUseError> {
        let input = serde_yaml::from_value(input)?;
        let output = self.invoke_typed(&input)?;
        Ok(serde_yaml::to_value(output)?)
    }
}
