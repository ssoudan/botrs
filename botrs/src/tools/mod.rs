use std::collections::HashMap;
use std::rc::Rc;

use llm_chain::tools::{Tool, ToolDescription, ToolUseError};

/// Hue tools
pub mod hue;

/// Tool to conclude a chain
pub mod conclude;

/// Tool to run some (limited) python
pub mod python;

#[cfg(test)]
/// Tool to test stuffs
pub(crate) mod dummy;

/// A termination message
pub struct TerminationMessage {
    /// The final textual answer for this task.
    pub conclusion: String,
    /// The original question that was asked to the user.
    pub original_question: String,
}

/// A tool that wraps a chain of exchanges
pub trait TerminalTool: Tool {
    /// done flag.
    fn is_done(&self) -> bool {
        false
    }

    /// Take the done flag.
    fn take_done(&self) -> Option<TerminationMessage> {
        None
    }
}

/// A tool that can benefit from a toolbox
pub trait AdvancedTool: Tool {
    /// Invoke the tool with a toolbox
    fn invoke_with_toolbox(
        &self,
        toolbox: Rc<Toolbox>,
        input: serde_yaml::Value,
    ) -> Result<serde_yaml::Value, ToolUseError>;
}

/// Toolbox
#[derive(Default)]
pub struct Toolbox {
    /// The terminal tools
    terminal_tools: HashMap<String, Box<dyn TerminalTool>>,

    /// The tools
    tools: HashMap<String, Box<dyn Tool>>,

    /// The advanced tools - the one that can invoke another tool (not an
    /// advanced one)
    advanced_tools: HashMap<String, Box<dyn AdvancedTool>>,
}

impl Toolbox {
    /// Collect the termination messages
    pub fn termination_messages(&self) -> Vec<TerminationMessage> {
        let mut messages = Vec::new();

        for tool in self.terminal_tools.values() {
            if let Some(message) = tool.take_done() {
                messages.push(message);
            }
        }

        messages
    }

    /// Add a terminal tool
    pub fn add_terminal_tool(&mut self, tool: impl TerminalTool + 'static) {
        let name = tool.description().name;
        self.terminal_tools.insert(name, Box::new(tool));
    }

    /// Add a tool
    pub fn add_tool(&mut self, tool: impl Tool + 'static) {
        let name = tool.description().name;
        self.tools.insert(name, Box::new(tool));
    }

    /// Add an advanced tool
    pub fn add_advanced_tool(&mut self, tool: impl AdvancedTool + 'static) {
        let name = tool.description().name;
        self.advanced_tools.insert(name, Box::new(tool));
    }

    /// Get the descriptions
    pub fn describe(&self) -> HashMap<String, ToolDescription> {
        let mut descriptions = HashMap::new();

        for (name, tool) in self.terminal_tools.iter() {
            descriptions.insert(name.clone(), tool.description());
        }

        for (name, tool) in self.tools.iter() {
            descriptions.insert(name.clone(), tool.description());
        }

        for (name, tool) in self.advanced_tools.iter() {
            descriptions.insert(name.clone(), tool.description());
        }

        descriptions
    }
}

/// Invoke a Tool or AdvancedTool  from a toolbox
pub fn invoke_from_toolbox(
    toolbox: Rc<Toolbox>,
    name: &str,
    input: serde_yaml::Value,
) -> Result<serde_yaml::Value, ToolUseError> {
    // test if the tool is an advanced tool
    if let Some(tool) = toolbox.clone().advanced_tools.get(name) {
        return tool.invoke_with_toolbox(toolbox, input);
    }

    // if not, test if the tool is a terminal tool
    if let Some(tool) = toolbox.terminal_tools.get(name) {
        return tool.invoke(input);
    }

    // otherwise, use the normal tool
    let tool = toolbox.tools.get(name).ok_or(ToolUseError::ToolNotFound)?;

    tool.invoke(input)
}

/// Invoke a Tool from a toolbox
pub fn invoke_simple_from_toolbox(
    toolbox: Rc<Toolbox>,
    name: &str,
    input: serde_yaml::Value,
) -> Result<serde_yaml::Value, ToolUseError> {
    // test if the tool is a terminal tool
    if let Some(tool) = toolbox.terminal_tools.get(name) {
        return tool.invoke(input);
    }

    // the normal tool only
    let tool = toolbox.tools.get(name).ok_or(ToolUseError::ToolNotFound)?;

    tool.invoke(input)
}
