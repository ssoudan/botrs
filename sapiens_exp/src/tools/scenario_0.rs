//! # Scenario 0
//! ----------
//! *"Make a cereal bowl" with "multiple ways to get to a single solution"*
//!
//! ## Sub-task
//! - Get bowl
//! - Get cereal
//! - Get milk
//! - Add cereal to bowl
//! - Add milk to bowl
//! - Serve cereal
//!
//! ## Tools
//! - Get <x>
//! - Add <x> to <y>
//! - Serve <x>
//!
//! ## With tools
//! - Get <bowl>
//! - Get <cereal>
//! - Get <milk>
//! - Add <cereal> to <bowl>
//! - Add <milk> to <bowl>
//! - Serve <bowl>
//!
//! ## Acceptance criteria
//! - The bowl contains the cereal &&
//! - The bowl contains the milk &&
//! - The bowl is served
//!
//! State machine:
//! - (no_bowl,no_cereal,no_milk)
//! - (no_bowl,cereal,no_milk)
//! - (no_bowl,cereal,milk)
//! - (bowl,no_cereal,milk)
//! - (bowl,cereal,no_milk)
//! - (bowl,no_cereal,no_milk)
//! - (bowl,cereal,no_milk)
//! - (bowl,cereal,milk)
//! - (served)

use std::sync::Arc;

use rust_fsm::*;
use sapiens::tools::toolbox::Toolbox;
use sapiens::tools::{Describe, ToolUseError};
use sapiens_derive::Describe;
use serde::{Deserialize, Serialize};
use tokio::sync::Mutex;

use crate::tools::{GenericTool, StateUpdater};

state_machine! {
    derive(Debug)
    repr_c(false)
    CerealBowlRecipe(NoBowlNoCerealNoMilk)

    // procurement
    NoBowlNoCerealNoMilk => {
        GetBowl => BowlNoCerealNoMilk [Found],
        GetCereal => NoBowlCerealNoMilk [Found],
        GetMilk => NoBowlNoCerealMilk [Found],
    },

    BowlNoCerealNoMilk => {
        GetCereal => BowlCerealNoMilk [Found],
        GetMilk => BowlNoCerealMilk [Found],
        GetBowl => BowlNoCerealNoMilk,
    },

    NoBowlCerealNoMilk => {
        GetBowl => BowlCerealNoMilk [Found],
        GetMilk => NoBowlCerealMilk [Found],
        GetCereal => NoBowlCerealNoMilk,
    },

    NoBowlNoCerealMilk => {
        GetBowl => BowlNoCerealMilk [Found],
        GetCereal => NoBowlNoCerealMilk [Found],
        GetMilk => NoBowlNoCerealMilk,
    },

    BowlCerealNoMilk => {
        GetBowl => BowlCerealNoMilk,
        GetCereal => BowlCerealNoMilk,
        GetMilk => BowlCerealMilk [Found],
    },

    BowlNoCerealMilk => {
        GetBowl => BowlNoCerealMilk,
        GetCereal => BowlCerealMilk [Found],
        GetMilk => BowlNoCerealMilk,
    },

    NoBowlCerealMilk => {
        GetBowl => BowlCerealMilk [Found],
        GetCereal => NoBowlCerealMilk,
        GetMilk => NoBowlCerealMilk,
    },

    // mixing
    BowlCerealMilk => {
        AddCerealToBowl => BowlWithCereal [Success],
        AddMilkToBowl => BowlWithMilk [Success],
    },
    BowlWithCereal(AddMilkToBowl) => BowlWithCerealAndMilk [Success],
    BowlWithMilk(AddCerealToBowl) => BowlWithCerealAndMilk [Success],

    // serving
    BowlWithCerealAndMilk(ServeBowl) => Served [Success],
}

/// The state of the scenario 0
pub struct State {
    fsm: StateMachine<CerealBowlRecipe>,
}

impl State {
    fn new() -> Self {
        Self {
            fsm: StateMachine::new(),
        }
    }

    /// is the scenario finished?
    pub fn has_reached_accepting_state(&self) -> bool {
        matches!(self.fsm.state(), CerealBowlRecipeState::Served)
    }

    /// Reset the state machine
    pub fn reset(&mut self) {
        self.fsm = StateMachine::new();
    }

    /// Get the current state
    #[allow(dead_code)]
    fn state(&self) -> &CerealBowlRecipeState {
        self.fsm.state()
    }
}

impl Default for State {
    fn default() -> Self {
        Self::new()
    }
}

/// What sits in the closet
#[derive(Debug, Serialize, Deserialize, Clone)]
enum ClosetObject {
    /// Bowl
    Bowl,
    /// Cereal
    Cereal,
    /// Milk
    Milk,
}

#[derive(Debug, Describe, Serialize, Deserialize, Clone)]
/// The input of a closet
struct ClosetInput {
    /// what to get: Bowl, Cereal, Milk
    get: ClosetObject,
}

impl StateUpdater<State, ClosetOutput> for ClosetInput {
    fn update(&self, state: &mut State) -> Result<ClosetOutput, ToolUseError> {
        match &self.get {
            ClosetObject::Bowl => {
                let x = state
                    .fsm
                    .consume(&CerealBowlRecipeInput::GetBowl)
                    .map_err(|e| ToolUseError::InvocationFailed(e.to_string()))?;
                Ok(x.into())
            }
            ClosetObject::Cereal => {
                let x = state
                    .fsm
                    .consume(&CerealBowlRecipeInput::GetCereal)
                    .map_err(|e| ToolUseError::InvocationFailed(e.to_string()))?;
                Ok(x.into())
            }
            ClosetObject::Milk => {
                let x = state
                    .fsm
                    .consume(&CerealBowlRecipeInput::GetMilk)
                    .map_err(|e| ToolUseError::InvocationFailed(e.to_string()))?;
                Ok(x.into())
            }
        }
    }
}

/// The output of a closet
#[derive(Debug, Serialize, Deserialize, Clone, Describe)]
struct ClosetOutput {
    /// was the object found?
    found: bool,
}

impl From<Option<CerealBowlRecipeOutput>> for ClosetOutput {
    fn from(output: Option<CerealBowlRecipeOutput>) -> Self {
        match output {
            Some(CerealBowlRecipeOutput::Found) => ClosetOutput { found: true },
            _ => ClosetOutput { found: false },
        }
    }
}

/// What to poor in
#[derive(Debug, Serialize, Deserialize, Clone)]
enum Container {
    /// Bowl
    Bowl,
}

/// What to pour
#[derive(Debug, Serialize, Deserialize, Clone)]
enum Pourable {
    /// Cereal
    Cereal,
    /// Milk
    Milk,
}

#[derive(Debug, Describe, Serialize, Deserialize, Clone)]
/// The input of a mixing action
struct MixingInput {
    /// what to pour in. Value can be: Bowl.
    container: Container,
    /// what to pour. Value can be: Cereal or Milk.
    pourable: Pourable,
}

impl StateUpdater<State, MixingOutput> for MixingInput {
    fn update(&self, state: &mut State) -> Result<MixingOutput, ToolUseError> {
        match (&self.container, &self.pourable) {
            (Container::Bowl, Pourable::Cereal) => {
                let x = state
                    .fsm
                    .consume(&CerealBowlRecipeInput::AddCerealToBowl)
                    .map_err(|e| ToolUseError::InvocationFailed(e.to_string()))?;
                Ok(x.into())
            }
            (Container::Bowl, Pourable::Milk) => {
                let x = state
                    .fsm
                    .consume(&CerealBowlRecipeInput::AddMilkToBowl)
                    .map_err(|e| ToolUseError::InvocationFailed(e.to_string()))?;
                Ok(x.into())
            }
        }
    }
}

/// The output of a mixing action
#[derive(Debug, Serialize, Deserialize, Clone, Describe)]
struct MixingOutput {
    /// was it poured?
    added: bool,
}

impl From<Option<CerealBowlRecipeOutput>> for MixingOutput {
    fn from(output: Option<CerealBowlRecipeOutput>) -> Self {
        match output {
            Some(CerealBowlRecipeOutput::Success) => MixingOutput { added: true },
            _ => MixingOutput { added: false },
        }
    }
}

/// What to serve
#[derive(Debug, Serialize, Deserialize, Clone)]
enum Serveable {
    /// Bowl
    Bowl,
}

#[derive(Debug, Describe, Serialize, Deserialize, Clone)]
/// The input of a serving action
struct ServingInput {
    /// what to serve. Value can be: bowl.
    serveable: Serveable,
}

impl StateUpdater<State, ServingOutput> for ServingInput {
    fn update(&self, state: &mut State) -> Result<ServingOutput, ToolUseError> {
        match &self.serveable {
            Serveable::Bowl => {
                let x = state
                    .fsm
                    .consume(&CerealBowlRecipeInput::ServeBowl)
                    .map_err(|e| ToolUseError::InvocationFailed(e.to_string()))?;
                Ok(x.into())
            }
        }
    }
}

/// The output of a serving action
#[derive(Debug, Serialize, Deserialize, Clone, Describe)]
struct ServingOutput {
    /// was it served?
    served: bool,
}

impl From<Option<CerealBowlRecipeOutput>> for ServingOutput {
    fn from(output: Option<CerealBowlRecipeOutput>) -> Self {
        match output {
            Some(CerealBowlRecipeOutput::Success) => ServingOutput { served: true },
            _ => ServingOutput { served: false },
        }
    }
}

/// Scenario 0
pub async fn build(mut toolbox: Toolbox) -> (Toolbox, Arc<Mutex<State>>) {
    let state = State::new();
    let shared_state = Arc::new(Mutex::new(state));

    let closet: GenericTool<ClosetInput, State, ClosetOutput> = GenericTool::new_with_descriptions(
        "closet".to_string(),
        "place where to find stuffs".to_string(),
        shared_state.clone(),
    );

    let mixing: GenericTool<MixingInput, State, MixingOutput> = GenericTool::new_with_descriptions(
        "mixing".to_string(),
        "when you need to mix things in a container".to_string(),
        shared_state.clone(),
    );

    let serving: GenericTool<ServingInput, State, ServingOutput> =
        GenericTool::new_with_descriptions(
            "serving".to_string(),
            "when the meal is ready to be served".to_string(),
            shared_state.clone(),
        );

    toolbox.add_tool(closet).await;
    toolbox.add_tool(mixing).await;
    toolbox.add_tool(serving).await;

    (toolbox, shared_state)
}

#[cfg(test)]
mod tests {
    use insta::assert_yaml_snapshot;
    use sapiens::tools::{ProtoToolInvoke, ToolDescription};

    use super::*;

    #[test]
    fn cereal_bowl() {
        let mut machine: StateMachine<CerealBowlRecipe> = StateMachine::new();
        println!("{:?}", machine.state());

        let _x = machine.consume(&CerealBowlRecipeInput::GetBowl).unwrap();
        println!("{:?}", machine.state());
        let _x = machine.consume(&CerealBowlRecipeInput::GetCereal).unwrap();
        println!("{:?}", machine.state());
        let _x = machine.consume(&CerealBowlRecipeInput::GetMilk).unwrap();
        let x = machine.state();
        println!("{:?}", x);

        let _x = machine
            .consume(&CerealBowlRecipeInput::AddCerealToBowl)
            .unwrap();
        println!("{:?}", machine.state());
        let _x = machine
            .consume(&CerealBowlRecipeInput::AddMilkToBowl)
            .unwrap();
        println!("{:?}", machine.state());
        let _x = machine.consume(&CerealBowlRecipeInput::ServeBowl).unwrap();
        println!("{:?}", machine.state());
    }

    #[tokio::test]
    async fn test_with_tools() {
        let shared_state = State::default();

        let shared_state = Arc::new(Mutex::new(shared_state));

        let closet: GenericTool<ClosetInput, State, ClosetOutput> =
            GenericTool::new_with_descriptions(
                "closet".to_string(),
                "place where to find stuffs".to_string(),
                shared_state.clone(),
            );

        // first get the bowl
        let input = ClosetInput {
            get: ClosetObject::Bowl,
        };

        let input = serde_yaml::to_value(input).unwrap();
        let output = closet.invoke(input).await.unwrap();
        let output = serde_yaml::from_value::<ClosetOutput>(output).unwrap();
        println!("{:?}", output);
        {
            let guard = shared_state.lock().await;
            let state = guard.state();
            if !matches!(state, CerealBowlRecipeState::BowlNoCerealNoMilk) {
                panic!("KO");
            }
        }

        // now get the cereal
        let input = ClosetInput {
            get: ClosetObject::Cereal,
        };

        let input = serde_yaml::to_value(input).unwrap();
        let output = closet.invoke(input).await.unwrap();
        let output = serde_yaml::from_value::<ClosetOutput>(output).unwrap();
        println!("{:?}", output);
        {
            let guard = shared_state.lock().await;
            let state = guard.state();
            if !matches!(state, CerealBowlRecipeState::BowlCerealNoMilk) {
                panic!("KO");
            }
        }

        // now get the milk
        let input = ClosetInput {
            get: ClosetObject::Milk,
        };

        let input = serde_yaml::to_value(input).unwrap();
        let output = closet.invoke(input).await.unwrap();
        let output = serde_yaml::from_value::<ClosetOutput>(output).unwrap();
        println!("{:?}", output);
        {
            let guard = shared_state.lock().await;
            let state = guard.state();
            if !matches!(state, CerealBowlRecipeState::BowlCerealMilk) {
                panic!("KO");
            }
        }

        // add a mixing tool
        let mixing: GenericTool<MixingInput, State, MixingOutput> =
            GenericTool::new_with_descriptions(
                "mixing".to_string(),
                "mixing stuffs".to_string(),
                shared_state.clone(),
            );

        // now mix the cereal
        let input = MixingInput {
            container: Container::Bowl,
            pourable: Pourable::Cereal,
        };

        let input = serde_yaml::to_value(input).unwrap();
        let output = mixing.invoke(input).await.unwrap();
        let output = serde_yaml::from_value::<MixingOutput>(output).unwrap();
        println!("{:?}", output);
        {
            let guard = shared_state.lock().await;
            let state = guard.state();
            if !matches!(state, CerealBowlRecipeState::BowlWithCereal) {
                panic!("KO");
            }
        }

        // now mix the milk
        let input = MixingInput {
            container: Container::Bowl,
            pourable: Pourable::Milk,
        };

        let input = serde_yaml::to_value(input).unwrap();
        let output = mixing.invoke(input).await.unwrap();
        let output = serde_yaml::from_value::<MixingOutput>(output).unwrap();
        println!("{:?}", output);
        {
            let guard = shared_state.lock().await;
            let state = guard.state();
            if !matches!(state, CerealBowlRecipeState::BowlWithCerealAndMilk) {
                panic!("KO");
            }
        }

        // add a serving tool
        let serving: GenericTool<ServingInput, State, ServingOutput> =
            GenericTool::new_with_descriptions(
                "serving".to_string(),
                "serving stuffs".to_string(),
                shared_state.clone(),
            );

        // now serve the bowl
        let input = ServingInput {
            serveable: Serveable::Bowl,
        };

        let input = serde_yaml::to_value(input).unwrap();
        let output = serving.invoke(input).await.unwrap();
        let output = serde_yaml::from_value::<ServingOutput>(output).unwrap();
        println!("{:?}", output);
        {
            let mut guard = shared_state.lock().await;
            let state = guard.state();
            if !matches!(state, CerealBowlRecipeState::Served) {
                panic!("KO");
            }

            assert!(
                guard.has_reached_accepting_state(),
                "Should be in accepting state"
            );

            // reset the state
            guard.reset();

            let state = guard.state();
            assert!(
                matches!(state, CerealBowlRecipeState::NoBowlNoCerealNoMilk),
                "Should be in initial state"
            );
        }

        // and we are done!
    }

    #[tokio::test]
    async fn test_with_toolbox() {
        let toolbox = Toolbox::default();

        let (toolbox, shared_state) = build(toolbox).await;

        {
            let guard = shared_state.lock().await;
            let state = guard.state();
            assert!(matches!(state, CerealBowlRecipeState::NoBowlNoCerealNoMilk));
        }

        let description = toolbox.describe().await;

        let mut description: Vec<ToolDescription> = description.into_values().collect();

        description.sort_by(|a, b| a.name.cmp(&b.name));

        assert_yaml_snapshot!(description);
    }
}
