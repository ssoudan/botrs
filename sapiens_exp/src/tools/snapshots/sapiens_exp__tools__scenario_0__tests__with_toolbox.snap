---
source: sapiens_exp/src/tools/scenario_0.rs
expression: description
---
- name: closet
  description: place where to find stuffs
  input_format:
    get: "<ClosetObject> what to get: Bowl, Cereal, Milk"
  output_format:
    found: "<bool> was the object found?"
- name: mixing
  description: when you need to mix things in a container
  input_format:
    container: "<Container> what to pour in. Value can be: Bowl."
    pourable: "<Pourable> what to pour. Value can be: Cereal or Milk."
  output_format:
    added: "<bool> was it poured?"
- name: serving
  description: when the meal is ready to be served
  input_format:
    serveable: "<Serveable> what to serve. Value can be: bowl."
  output_format:
    served: "<bool> was it served?"

